File: cpp/2899.last-visited-integers.cpp
#include <vector>

using std::vector;

class Solution {
public:
  vector<int> lastVisitedIntegers(vector<int> &nums) {
    vector<int> ans, seen;
    int k = 0;

    for (int num : nums) {
      if (num > 0) {
        seen.push_back(num);
        k = 0;
      } else {
        if (seen.size() - k > 0) {
          ans.push_back(seen[seen.size() - ++k]);
        } else {
          ans.push_back(-1);
        }
      }
    }

    return ans;
  }
};

File: golang/2899.last-visited-integers.go
func lastVisitedIntegers(nums []int) []int {
	k := 0
	ans, seen := make([]int, 0), make([]int, 0)

	for _, num := range nums {
		if num > 0 {
			seen = append(seen, num)
			k = 0
		} else {
			if len(seen)-k > 0 {
				ans = append(ans, seen[len(seen)-k-1])
				k++
			} else {
				ans = append(ans, -1)
			}
		}
	}

	return ans
}
